Fix high frag in gen2

PeterSolMS opened the pull request:
Details:
 - New config setting GCConserveMemory with values between 0 and 9 - intuitively, the higher the value, the more GC tries to keep gen 2 compact.
- Implementation: the config value is considered in dt_high_frag_p as the desired percentage/10 of useful data in gen 2. In other words, 100 - config value*10 is the fragmentation threshold above which a blocking gen 2 is performed. For example, if the config setting is 3, we do a blocking gen 2 if the fragmentation in gen 2 is higher than 70%.

Initial perf data suggest this does have the desired effect of trimming down average gen 2 size (and especially trimming it down when the survivorship in gen 2 goes down). There was also the expected effect of larger GC pause times, and an unexpected effect of larger LOH sizes, probably because we also trigger fewer gen 2 collections.

PeterSolMS committed:
delete backport.yml so I can push to master

PeterSolMS committed:
delete backport.yml so I can push to master

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Undo unintended changes

PeterSolMS committed:
Undo unintended changes

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge from upstream/master

PeterSolMS committed:
Merge from upstream/master

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge branch 'master' of https://github.com/PeterSolMS/runtime-1

PeterSolMS committed:
Undo unintended Mono changes that somehow crept in.

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Introduce a config setting that triggers a compacting gen 2 collection when gen 2 fragmentation becomes too high.

Details:
 - New config setting GCConserveMemory with values between 0 and 9 - intuitively, the higher the value, the more GC tries to keep gen 2 compact.
- Implementation: the config value is considered in dt_high_frag_p as the desired percentage*10 of useful data in gen 2. In other words, 100 - config value*10 is the fragmentation threshold above which a blocking gen 2 is performed.

mangod9 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -2989,17 +2989,25 @@ gc_heap::dt_high_frag_p (gc_tuning_point tp,
+                        conserve_mem_setting = 0;
+                    if (conserve_mem_setting > 9)
+                        conserve_mem_setting = 9;
+                    float frag_limit = 1.0f - conserve_mem_setting / 10.0f;
```
assume this is a customer facing setting? Would it be easier to document if it was represented as straight percent, instead of between 0-9?

Maoni0 replied:
we specifically did not want this to be a particular internal metric in GC and instead wanted a dial so GC can change the implementation as needed.

PeterSolMS replied:
I consider the implementation as an internal detail. 

The intuitive meaning is simply: the higher you set it, the harder GC will try to conserve memory, at the expense of using more time in GC pauses. Plus, I didn't want this setting to imply any kind of precision - people shouldn't be tempted to complain if they set it to 54% and we use 57%.

mangod9 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -2989,17 +2989,25 @@ gc_heap::dt_high_frag_p (gc_tuning_point tp,
+                    float frag_limit = 1.0f - conserve_mem_setting / 10.0f;
+#ifndef MULTIPLE_HEAPS
+                    if (conserve_mem_setting == 0)
+                        frag_limit = 0.65f;
```
is there a different default for `MULTIPLE_HEAPS`?

PeterSolMS replied:
The default for MULTIPLE_HEAPS is 1.0, i.e. no limit. This was the behavior before, I didn't want to change it.

PeterSolMS committed:
Add more sophistication to the conserve memory logic:
- store the config setting in a static field on gc_heap, fetch & validate config in init_semi_shared.
- take LOH fragmentation into account as well
- trim down budgets for gen 2 and above if conserve memory setting is aggressive enough

PeterSolMS committed:
Added dprintf for the setting of GCConserveMemory.

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge branch 'master' into Fix_high_frag_in_gen2

PeterSolMS committed:
Fixed bitrot with FREE_USAGE_STATS, additional debug output.

PeterSolMS committed:
Merge remote-tracking branch 'upstream/main' into main

PeterSolMS committed:
Disable mechanism to reduce budget for gen 2 if conserve_mem_setting is on.

PeterSolMS committed:
Merge branch 'main' into Fix_high_frag_in_gen2

PeterSolMS committed:
Merge remote-tracking branch 'upstream/main' into main

PeterSolMS committed:
Update logic to look at the combined fragmentation in gen 2 and LOH to make the compaction decision. Also, only do the computations only for gen 1 GCs to cut down on the back-to-back gen 2 GCs that we were observing.

PeterSolMS committed:
Merge remote-tracking branch 'upstream/main' into main

PeterSolMS committed:
Merge remote-tracking branch 'upstream/main' into main

PeterSolMS committed:
Merge remote-tracking branch 'upstream/main' into main

PeterSolMS committed:
Merge remote-tracking branch 'upstream/main' into main

PeterSolMS committed:
Merge branch 'main' into Fix_high_frag_in_gen2

PeterSolMS committed:
Base decision on fully compacting GC in the hard limit case on the combined fragmentation in gen 2 + loh, rather than just the fragmentation in loh.

PeterSolMS committed:
Merge with main

PeterSolMS committed:
Fix dprintfs that had bitrotted and were commented out, turned off TRACE_GC and removed debugging code.

PeterSolMS committed:
Remove extra file.
